src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterBuilder.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Core;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    /// <summary>
+    /// Builds a middleware pipeline after receiving the pipeline from a pipeline provider
+    /// </summary>
+    public class MiddlewareFilterBuilder
+    {
+        private readonly ConcurrentDictionary<Type, Lazy<RequestDelegate>> _pipelinesCache
```

Comment:
The use of Lazy here really need to be documented well. It's totally non obvious and also important